File: src/utils/log.cpp
#include <quart/utils/log.h>

#include <algorithm>
#include <sstream>

std::string format_location_span(const Span& span) {
    return FORMAT("{0}:{1}:{2}", span.filename, span.start.line, span.start.column);
}

bool utils::has_color_support() {
    return isatty(fileno(stdout));
}

std::string utils::color_to_str(Color color) {
    return FORMAT("\033[1;{0}m", int(color));
}

std::string utils::color(Color color, const std::string& str) {
    if (utils::has_color_support()) {
        return FORMAT("\033[1;{0}m{1}\033[0;0m", int(color), str);
    } else {
        return str;
    }
}

void utils::error(const Span& span, const std::string& message, bool fatal) {
    std::stringstream stream;

    stream << utils::color(WHITE, format_location_span(span)) << ' ';
    stream << utils::color(RED, "error:") << ' ';

    stream << message << '\n';

    std::string fmt = FORMAT("{0} | ", span.start.line);
    stream << utils::color(WHITE, fmt) << span.line << '\n';

    size_t len = span.line.length();
    size_t offset = std::min(span.start.column - 1 + fmt.length(), len);

    uint32_t padding = span.length() <= 0 ? 1 : span.length();
    for (size_t i = 0; i < offset; i++) stream << ' ';
    for (size_t i = 0; i < padding; i++) stream << '^';

    stream << '\n';

    std::cout << stream.str() << std::endl;
    if (fatal) {
        exit(1);
    }
}

void utils::note(const Span& span, const std::string& message) {
    std::stringstream stream;

    stream << utils::color(WHITE, format_location_span(span)) << ' ';
    stream << utils::color(MAGENTA, "note:") << ' ';

    stream << message << '\n';

    std::string fmt = FORMAT("{0} | ", span.start.line);
    stream << utils::color(WHITE, fmt) << span.line << '\n';

    size_t len = span.line.length();
    size_t offset = std::min(span.start.column - 1 + fmt.length(), len);

    for (size_t i = 0; i < offset; i++) stream << ' ';
    for (size_t i = 0; i < span.length(); i++) stream << '^';

    stream << '\n';
    std::cout << stream.str() << std::endl;
}

File: rust/src/log.rs
use crate::tokens::Span;

use colored::*;

fn format_span(span: &Span) -> String {
    format!("{}:{}:{}", span.filename, span.start.line, span.start.column)
}

// TODO: Disable colors when not in a TTY
pub fn log_error(span: &Span, message: String) {
    println!("{} {} {}", format_span(span).bold(), "error:".red().bold(), message);

    let fmt = format!("{} |", span.start.line);
    println!("{} {}", fmt.bold(), span.line);
    
    let len = span.line.len();
    let offset = std::cmp::min(
        span.start.column - 1 + fmt.len(), len
    );

    let padding = if span.length() == 0 { 1 } else { span.length() };
    for _ in 0..offset {
        print!(" ");
    }

    for _ in 0..padding {
        print!("{}", "^".bold());
    }

    println!();
}

pub fn log_note(span: &Span, message: String) {
    println!("{} {} {}", format_span(span).bold(), "note:".purple().bold(), message);

    let fmt = format!("{} |", span.start.line);
    println!("{} {}", fmt.bold(), span.line);
    
    let len = span.line.len();
    let offset = std::cmp::min(
        span.start.column - 1 + fmt.len(), len
    );

    let padding = if span.length() == 0 { 1 } else { span.length() };
    for _ in 0..offset {
        print!(" ");
    }

    for _ in 0..padding {
        print!("{}", "^".bold());
    }

    println!();
}

pub fn default_log_error(message: String) {
    println!("{} {} {}", "quart:".bold(), "error:".red().bold(), message);
}

macro_rules! error {
    ($message:literal) => {
        { 
            use crate::log::default_log_error;

            default_log_error($message.to_string());
            std::process::exit(1);
        }
    };

    ($message:literal, $($arg:tt)*) => {
        {
            use crate::log::default_log_error;

            default_log_error(format!($message, $($arg)*));
            std::process::exit(1);
        }
    };

    ($span:expr, $message:expr) => {
        {
            use crate::log::log_error;

            log_error($span, $message.to_string());
            std::process::exit(1);
        }
    };

    ($span:expr, $message:expr, $($arg:tt)*) => {
        {
            use crate::log::log_error;

            log_error($span, format!($message, $($arg)*));
            std::process::exit(1);
        }
    };
}

macro_rules! note {
    ($span:expr, $message:expr) => {
        {
            use crate::log::log_note;
            log_note($span, $message.to_string());
        }
    };

    ($span:expr, $message:expr, $($arg:tt)*) => {
        {
            use crate::log::log_note;
            log_note($span, format!($message, $($arg)*));
        }
    };
}

pub(crate) use error;
pub(crate) use note;